Rami4atm experiment

schunkes opened the pull request:
# Description

This PR adds the RAMI4ATM experiment and a specialized surface based on the `blendbsdf` mitsuba plugin.

Once the first round of feedback is done I will add more elaborate tests that compare the Rami4ATMExperiment's edge cases to the OneeDimExperiment and RamiExperiment classes.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

leroyvn commented on eradiate/experiments/_rami4atm.py:
```diff
@@ -0,0 +1,197 @@
+        result = super(Rami4ATMExperiment, self)._dataset_metadata(measure)
+
+        if isinstance(measure, DistantMeasure):
+            result["title"] = "Top-of-canopy simulation results"
```
```suggestion
            result["title"] = "Top-of-atmosphere simulation results"
```

leroyvn commented on eradiate/experiments/_rami4atm.py:
```diff
@@ -0,0 +1,197 @@
+        default=":class:`VolPathIntegrator() <.VolPathIntegrator>`",
+    )
+
+    def __attrs_pre_init__(self):
+        # Only tested with monochromatic modes
+        supported_mode(ModeFlags.ANY_MONO)
```
```suggestion
```

schunkes commented:
Alright, here we go again. The docstring for the CentralPatchSurface might be a bit too short, but I wasn't sure how to describe it properly.

leroyvn commented on eradiate/contexts.py:
```diff
@@ -300,3 +300,18 @@ class KernelDictContext(Context):
+        "has to match a padded canopy).\n"
+        "\n"
+        "Unit-enabled field (default: ucc['length']).",
+        type="quantity, optional",
```
```suggestion
        type="quantity or None",
```

leroyvn commented on eradiate/experiments/_rami4atm.py:
```diff
@@ -0,0 +1,200 @@
+
+    Notes
+    -----
+    A post-initialisation step will constrain the measure setup if a
+    :class:`.DistantMeasure` is used and no target is defined:
+
+    * if a canopy is defined, the target will be set to the canopy unit cell
+      (*i.e.* without its padding);
+    * if no canopy is defined, the target will be set to [0, 0, 0].
```
Doesn't this need an update?

leroyvn commented on eradiate/scenes/surface/_centralpatch.py:
```diff
@@ -0,0 +1,129 @@
+        "constraint (*e.g.* to match the size of an atmosphere or canopy) is applied.\n"
+        "\n"
+        "Unit-enabled field (default: ucc['length']).",
+        type="float or AUTO, optional",
```
```suggestion
        type="float or AUTO",
        init_type="float or AUTO, optional",
```

leroyvn commented on eradiate/scenes/surface/_centralpatch.py:
```diff
@@ -0,0 +1,129 @@
+    def bsdfs(self, ctx: KernelDictContext) -> KernelDict:
+        from mitsuba.core import ScalarTransform4f
+
+        returndict = {
```
This should be a `KernelDict`.

leroyvn commented on eradiate/scenes/surface/_centralpatch.py:
```diff
@@ -0,0 +1,129 @@
+    """
+    Central patch surface element [``central_patch``]
+
+    This class creates a square surface to which two BSDFs will be attached.
```
That's a bit too vague, would you mind elaborating a bit?

leroyvn commented on eradiate/scenes/surface/_centralpatch.py:
```diff
@@ -0,0 +1,129 @@
+            converter=surface_factory.convert,
+            validator=attr.validators.instance_of(Surface),
+        ),
+        doc="Inner surface specification. "
```
```suggestion
        doc="Central patch specification. "
```

leroyvn commented on eradiate/scenes/surface/_centralpatch.py:
```diff
@@ -0,0 +1,129 @@
+        default=":class:`LambertianSurface() <.LambertianSurface>`",
+    )
+
+    outer_surface: Surface = documented(
```
How about `background_surface` instead, to clearly emphasise the idea of a background on which a patch is applied.

leroyvn commented on eradiate/scenes/surface/_centralpatch.py:
```diff
@@ -0,0 +1,129 @@
+        default=":class:`LambertianSurface() <.LambertianSurface>`",
+    )
+
+    patch_width: t.Union[pint.Quantity, AutoType] = documented(
```
The more I think about it, the more I believe that this parameter can be removed. You can just use `central_patch.width` instead. What do you think?

leroyvn reviewed (changes requested):
Hi @schunkes, we're almost there. Just a couple of small changes and we'll be good to go!

leroyvn commented on eradiate/scenes/surface/_centralpatch.py:
```diff
@@ -0,0 +1,136 @@
+        "\n"
+        "Unit-enabled field (default: ucc['length']).",
+        type="float or AUTO",
+        default="float or AUTO, optional",
```
```suggestion
        init_type="float or AUTO, optional",
        default="AUTO",
```

schunkes commented:
I implemented the things we discussed yesterday. 

Additionally I added a separate commit, where I set the default ray origin to cover the entire horizontal extent of the scene and be located a small distance above either the surface, canopy or atmosphere, whichever is highest.
For that I had to allow the ray origin to be set to `TargetOriginRectangle` and I don't know if there was a reason for not allowing that, which is why I kept this as a separate commit and we can leave it out if you think it should not be used.
